File: p122e/max_profit_inf.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        ans = 0
        for i in range(len(prices) - 1):
            ans += max(prices[i + 1] - prices[i], 0)
        return ans


# TESTS
tests = [
    [[1], 0],
    [[7, 1, 5, 3, 6, 4], 7],
    [[1, 2, 3, 4, 5], 4],
    [[7, 6, 4, 3, 1], 0],
]
for t in tests:
    sol = Solution()
    actual = sol.maxProfit(t[0])
    print("Max profix by inf transaction in", t[0], "->", actual)
    assert actual == t[1]

File: p122e/max_profit_inf.go
package main

import "fmt"

func maxProfit(prices []int) int {
	ans := 0
	for i := 1; i < len(prices); i++ {
		if prices[i] > prices[i-1] {
			ans += prices[i] - prices[i-1]
		}
	}
	return ans
}

func main() {
	tests := [][]int{
		{},
		{1},
		{7, 1, 5, 3, 6, 4},
		{7, 6, 4, 3, 1},
	}
	for _, t := range tests {
		fmt.Println("Max profit by inf transaction in", t,
			"->", maxProfit(t))
	}
}
